source/common/network/connection_impl.cc

```diff
@@ -97,6 +97,11 @@ void ConnectionImpl::close(ConnectionCloseType type) {
     return;
   }
 
+  // We're now closed, there's no reason to read from this point forward. This
+  // is important to set explicitly, to handle things like delayed close. This
+  // will also apply TCP back pressure.
+  read_enabled_ = false;
```

Comment:
In looking at this code a little more, I'm still a little worried about the fragility of this, especially how this interacts with `readDisable()`. Should we actually be checking close state flags inside `onRead()` and not interacting with this variable at all here and instead just using it to track the state of the `readDisable()` call? @alyssawilk @AndresGuedez WDYT?

Agree with @alyssawilk also that we should definitely have an additional unit test for this case and any other coverage we can add.

Reply:
+1 to not setting `read_enabled_ = false` in `close()`.

Checking `inDelayedClose()`, `read_enabled_` and possibly also asserting `ioHandle().isOpen()` in `onRead()` should provide enough protection.